daemon/api_prompting.go

```diff
@@ -97,6 +99,92 @@ func getUserID(r *http.Request) (uint32, Response) {
 }
 
+func promptingNotRunningError() *apiError {
+	return &apiError{
+		Status:  500, // Internal error
+		Message: "AppArmor Prompting is not running",
+		Kind:    client.ErrorKindAppArmorPromptingNotRunning,
+	}
+}
+
+func promptingError(err error) *apiError {
+	apiErr := &apiError{
+		Message: err.Error(),
+	}
+	switch {
+	case errorIsOneOf(err, prompting.ErrPromptsClosed, prompting.ErrRulesClosed):
+		apiErr.Status = 500
+		apiErr.Kind = client.ErrorKindInterfacesRequestsManagerClosed
+	case errors.Is(err, prompting.ErrInvalidID):
+		apiErr.Status = 400
+		apiErr.Kind = client.ErrorKindInterfacesRequestsInvalidID
+	case errors.Is(err, prompting.ErrPromptNotFound):
+		apiErr.Status = 404
+		apiErr.Kind = client.ErrorKindInterfacesRequestsPromptNotFound
+	case errorIsOneOf(err, prompting.ErrRuleNotFound, prompting.ErrRuleNotAllowed):
+		// Even if the error is ErrRuleNotAllowed, reply with 404 status
+		// to match the behavior of prompts, and so if we switch to storing
+		// rules by ID (and don't want to check whether a rule with that ID
+		// exists for some other user), this error will remain unchanged.
+		apiErr.Status = 404
+		apiErr.Kind = client.ErrorKindInterfacesRequestsRuleNotFound
+	case errors.Is(err, prompting.ErrInvalidOutcome):
+		apiErr.Status = 400
+		apiErr.Kind = client.ErrorKindInterfacesRequestsInvalidOutcome
+	case errors.Is(err, prompting.ErrInvalidLifespan):
+		apiErr.Status = 400
+		apiErr.Kind = client.ErrorKindInterfacesRequestsInvalidLifespan
+	case errors.Is(err, prompting.ErrInvalidDuration):
+		apiErr.Status = 400
+		apiErr.Kind = client.ErrorKindInterfacesRequestsInvalidDuration
+	case errors.Is(err, prompting.ErrInvalidExpiration):
+		apiErr.Status = 400
+		apiErr.Kind = client.ErrorKindInterfacesRequestsInvalidExpiration
+	case errors.Is(err, prompting.ErrInvalidConstraints):
+		apiErr.Status = 400
+		apiErr.Kind = client.ErrorKindInterfacesRequestsInvalidConstraints
+	case errors.Is(err, prompting.ErrRuleExpirationInThePast):
+		apiErr.Status = 400
+		apiErr.Kind = client.ErrorKindInterfacesRequestsRuleExpirationInThePast
+	case errors.Is(err, prompting.ErrRuleLifespanSingle):
+		apiErr.Status = 400
+		apiErr.Kind = client.ErrorKindInterfacesRequestsInvalidDuration
+	case errors.Is(err, prompting.ErrReplyNotMatchRequestedPath):
+		apiErr.Status = 400
+		apiErr.Kind = client.ErrorKindInterfacesRequestsReplyNotMatchRequestedPath
+	case errors.Is(err, prompting.ErrReplyNotMatchRequestedPermissions):
+		apiErr.Status = 400
+		apiErr.Kind = client.ErrorKindInterfacesRequestsReplyNotMatchRequestedPermissions
+	case errors.Is(err, prompting.ErrRuleConflict):
+		apiErr.Status = 409
+		apiErr.Kind = client.ErrorKindInterfacesRequestsRuleConflict
+		var conflictErr *requestrules.RuleConflictError
+		if errors.As(err, &conflictErr) {
+			// Only include RuleConflict message, as the Value can hold the
+			// conflict details
+			apiErr.Message = prompting.ErrRuleConflict.Error()
+			apiErr.Value = conflictErr
```

Comment:
isn't this leaking package implementation detail?

Reply:
Kind of? It's leaking error implementation details at least, with the fact that we expect `ErrRuleConflict` errors to be wrapped in `RuleConflictError`s. The error itself is designed to be marshalled in this way, so I don't think details beyond that fact are leaked by putting it in the Value.